crates/external/rpc/src/traits.rs

```diff
@@ -51,16 +64,48 @@ pub trait ExecutingMessageValidator {
         let timeout = timeout.map_or(Self::DEFAULT_TIMEOUT, |t| t);
 
         // Construct the future to validate all messages using supervisor.
-        let fut = async {
-            supervisor
-                .check_messages(messages.to_vec(), safety)
-                .await
-                .map_err(ExecutingMessageValidatorError::SupervisorRpcError)
-        };
+        let fut = async { supervisor.check_messages(messages.to_vec(), safety).await };
 
         // Await the validation future with timeout.
-        tokio::time::timeout(timeout, fut)
+        let res = tokio::time::timeout(timeout, fut)
             .await
-            .map_err(ExecutingMessageValidatorError::ValidationTimeout)?
+            .map_err(ExecutingMessageValidatorError::ValidationTimeout)?;
+        match res {
+            Ok(_) => Ok(()),
+            Err(err) => {
+                match err {
+                    ClientError::Call(err) => {
+                        // Check if it's invalid message call, message example:
+                        // `failed to check message: failed to check log: unknown chain: 14417`
+                        if let Some(msg_start) = err.message().find(UNKNOWN_CHAIN_MSG) {
+                            let chain_id_start = msg_start + UNKNOWN_CHAIN_MSG.len();
+                            if let Ok(chain_id) = &err.message()[chain_id_start..].parse::<u64>() {
+                                return Err(ExecutingMessageValidatorError::UnknownChain(*chain_id));
+                            }
+                        // Check if it's `does not meet the minimum safety` error, message example:
+                        // `message {0x4200000000000000000000000000000000000023 4 1 1728507701 901}
+                        // (safety level: unsafe) does not meet the minimum safety cross-unsafe"`
+                        } else if err.message().contains(MINIMUM_SAFETY_MSG) {
+                            if let Some(msg_start) = err.message().find(SAFETY_START_MSG) {
+                                let safety_start = msg_start + SAFETY_START_MSG.len();
+                                if let Some(safety_end) = &err.message()[safety_start..].find(")") {
+                                    if let Ok(got_safety_level) = (&err.message()
+                                        [safety_start..safety_start + safety_end])
+                                        .try_into()
+                                    {
+                                        return Err(ExecutingMessageValidatorError::MinimumSafety(
+                                            safety,
+                                            got_safety_level,
+                                        ));
+                                    }
+                                }
+                            }
+                        }
```

Comment:
think it's cleaner here to just match on the literals, i.e.
```rust
let got = if err.message().contains("safety level: safe") {
    SafetyLevel::Safe
} else if err.message().contains("safety level: local-safe") {
    SafetyLevel::LocalSafe
} else if ..;
```
and getting minimum with `err.message().split(' ').last().expect("non-empty string")`